### PR TITLE
WS-037: Add, maintain and delete species functions and pages for admins

### PR DESCRIPTION
**Pull request by AliThomson**
image
image
image
image

### CODE REVIEW

File: backend/Controllers/SpeciesController.cs [hunk 1]
```diff
@@ -14,20 +15,207 @@ namespace WhaleSpotting.Controllers
...
+
+            catch (InvalidOperationException)
+            {
+                return StatusCode(
+                    StatusCodes.Status401Unauthorized,
+                    "The given username is not valid"
+                );
+            }
+            
+            if (user.Role == 0)
```
Comment (TimLeach635) [inline backend/Controllers/SpeciesController.cs:132]: Best to check if the user is _not_ an admin - if we later add a new role, like "Guest", they would by default be allowed to access this endpoint!

File: backend/Controllers/SpeciesController.cs [hunk 2]
```diff
@@ -14,20 +15,207 @@ namespace WhaleSpotting.Controllers
...
```
Comment (TimLeach635) [inline backend/Controllers/SpeciesController.cs:69]: Comment below

Comment (TimLeach635) [inline backend/Controllers/SpeciesController.cs:69]: Comment below

File: backend/Controllers/SpeciesController.cs [hunk 3]
```diff
@@ -14,20 +15,207 @@ namespace WhaleSpotting.Controllers
...
+            }
+
+            var check = _authservice.IsAuthenticated(usernamePassword);
+            
+            if (!check)
+            {
+                return new UnauthorizedResult();
+            }
+            
+             if (user.Role == 0)
```
Comment (TimLeach635) [inline backend/Controllers/SpeciesController.cs:207]: Likewise above

Comment (TimLeach635) [inline backend/Controllers/SpeciesController.cs:207]: Likewise above

File: backend/Models/Request/CreateSpeciesRequest.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            Name = species.Name;
+            LatinName = species.LatinName;
+            Description = species.Description;
+            PhotoUrl = species.PhotoUrl;
+            EndangeredStatusId = species.EndangeredStatusId;
+        }
+        public CreateSpeciesRequest ()
+        {
+        }
```
Comment (TimLeach635) [inline backend/Models/Request/CreateSpeciesRequest.cs:32]: Don't change this, but you'll never use this constructor - this class gets parsed from HTTP requests sent to this API, so you will never be creating an instance of this class in code, only ever reading and using it!

Comment (TimLeach635) [inline backend/Models/Request/CreateSpeciesRequest.cs:32]: Don't change this, but you'll never use this constructor - this class gets parsed from HTTP requests sent to this API, so you will never be creating an instance of this class in code, only ever reading and using it!

File: backend/Models/Response/UpdatedSpeciesResponse.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        public int EndangeredStatusId { get; set; }
+        public UpdatedSpeciesResponse(Species species)
+        {
+            Id = species.Id;
+            Name = species.Name;
+            LatinName = species.LatinName;
+            Description = species.Description;
+            PhotoUrl = species.PhotoUrl;
+            EndangeredStatusId = species.EndangeredStatusId;
+        }
```
Comment (TimLeach635) [inline backend/Models/Response/UpdatedSpeciesResponse.cs:21]: Likewise above (but you don't need to change it)

Comment (TimLeach635) [inline backend/Models/Response/UpdatedSpeciesResponse.cs:21]: Likewise above (but you don't need to change it)

File: backend/Repositories/SpeciesRepo.cs [hunk 6]
```diff
@@ -25,7 +30,51 @@ public List<Species> GetAllSpecies()
...
+        {
+            var updatedResult = GetSpeciesById(speciesId);
+            
+            updatedResult.Name = speciesToUpdate.Name;
+            updatedResult.LatinName = speciesToUpdate.LatinName;
+            updatedResult.PhotoUrl = speciesToUpdate.PhotoUrl;
+            updatedResult.Description = speciesToUpdate.Description;
+            updatedResult.EndangeredStatusId = speciesToUpdate.EndangeredStatusId;
+        
+            _context.Species.Update(updatedResult);
```
Comment (TimLeach635) [inline backend/Repositories/SpeciesRepo.cs:69]: I have a _slight_ suspicion that you don't need this line... remove it and test it and see!

Comment (TimLeach635) [inline backend/Repositories/SpeciesRepo.cs:69]: I have a _slight_ suspicion that you don't need this line... remove it and test it and see!

File: frontend/src/clients/apiClients.ts [hunk 7]
```diff
@@ -12,11 +12,11 @@ export interface NewUser {
...
-export interface Species {
-  description: string;
-  name: string;
-  latinName: string;
-}
+// export interface Species {
+//   description: string;
+//   name: string;
+//   latinName: string;
+// }
```
Comment (TimLeach635) [inline frontend/src/clients/apiClients.ts:19]: Remove comments!

Comment (TimLeach635) [inline frontend/src/clients/apiClients.ts:19]: Remove comments!

File: frontend/src/clients/apiClients.ts [hunk 8]
```diff
@@ -238,3 +267,65 @@ export const getLeaderboard = async () => {
   }
   return data;
 };
+
+export async function fetchEndangeredStatus(): Promise<
+  Array<EndangeredStatus>
+> {
+  const response = await fetch(`https://localhost:5001/endangered`);
```
Comment (TimLeach635) [inline frontend/src/clients/apiClients.ts:274]: You'll need to merge and change this to use the `baseUrl` variable

Comment (TimLeach635) [inline frontend/src/clients/apiClients.ts:274]: You'll need to merge and change this to use the `baseUrl` variable

File: frontend/src/clients/apiClients.ts [hunk 9]
```diff
@@ -238,3 +267,65 @@ export const getLeaderboard = async () => {
...
+  return await response.json();
+}
+
+export async function createSpecies(
+  newSpecies: NewSpecies,
+  username: string,
+  password: string
+) {
+  const response = await fetch(`https://localhost:5001/species/create`, {
```
Comment (TimLeach635) [inline frontend/src/clients/apiClients.ts:283]: ditto

Comment (TimLeach635) [inline frontend/src/clients/apiClients.ts:283]: ditto

File: frontend/src/clients/apiClients.ts [hunk 10]
```diff
@@ -238,3 +267,65 @@ export const getLeaderboard = async () => {
...
+  }
+}
+
+export async function updateSpecies(
+  id: number,
+  updatedSpecies: UpdateSpecies,
+  username: string,
+  password: string
+) {
+  const response = await fetch(`https://localhost:5001/species/${id}/update`, {
```
Comment (TimLeach635) [inline frontend/src/clients/apiClients.ts:302]: ditto

Comment (TimLeach635) [inline frontend/src/clients/apiClients.ts:302]: ditto

File: frontend/src/clients/apiClients.ts [hunk 11]
```diff
@@ -238,3 +267,65 @@ export const getLeaderboard = async () => {
...
+    throw new Error(await response.json());
+  }
+}
+
+export async function deleteSpecies(
+  id: number,
+  username: string,
+  password: string
+) {
+  const response = await fetch(`https://localhost:5001/species/${id}`, {
```
Comment (TimLeach635) [inline frontend/src/clients/apiClients.ts:321]: ditto

Comment (TimLeach635) [inline frontend/src/clients/apiClients.ts:321]: ditto

**Review by TimLeach635 — CHANGES_REQUESTED** [review]
Just a few comments!

File: backend/Controllers/SpeciesController.cs (hunk 1 above)
Comment (AliThomson) [reply backend/Controllers/SpeciesController.cs:132]: Done

File: frontend/src/clients/apiClients.ts (hunk 7 above)
Comment (AliThomson) [reply frontend/src/clients/apiClients.ts:19]: Done

Comment (AliThomson) [reply frontend/src/clients/apiClients.ts:19]: Done

File: backend/Repositories/SpeciesRepo.cs (hunk 6 above)
Comment (AliThomson) [reply backend/Repositories/SpeciesRepo.cs]: done

**Comment by AliThomson** [comment]
All requested changes made, ready for rereview when you get a minute!
Thanks
Alison and Ling

**Review by TimLeach635 — APPROVED** [review]

**Pull request merged**